here-naksha-lib-psql/src/commonMain/kotlin/naksha/psql/executors/write/UpdateFeature.kt

```diff
@@ -0,0 +1,206 @@
+import naksha.model.Version
+import naksha.model.objects.NakshaFeature
+import naksha.model.request.Write
+import naksha.psql.PgCollection
+import naksha.psql.PgColumn
+import naksha.psql.PgCursor
+import naksha.psql.PgSession
+import naksha.psql.PgUtil.PgUtilCompanion.quoteIdent
+import naksha.psql.executors.write.WriteFeatureUtils.newFeatureTupleNumber
+import naksha.psql.executors.write.WriteFeatureUtils.resolveFlags
+import naksha.psql.executors.write.WriteFeatureUtils.tuple
+
+class UpdateFeature(
+    val session: PgSession
+) {
+
+    fun execute(collection: PgCollection, write: Write): Tuple {
+        val feature = write.feature?.proxy(NakshaFeature::class) ?: throw NakshaException(
+            NakshaError.ILLEGAL_ARGUMENT,
+            "UPDATE without feature"
+        )
+        require(feature.id == write.featureId) {
+            "Feature id in payload (${feature.id}) and write request (${write.featureId}) are different"
+        }
+
+        val cursorWithFeatureHead = fetchFromHead(collection, feature.id)
+        val previousMetadata = metaFromCursor(cursorWithFeatureHead)
+
+        require(feature.id == previousMetadata.id) {
+            "Feature id (${feature.id}) differs from previous metadata (${previousMetadata.id})"
+        }
+        require(previousMetadata.nextVersion == null) {
+            "Previous metadata shouldn't have 'nextVersion' but it does (${previousMetadata.nextVersion})"
+        }
+
+        val tupleNumber = newFeatureTupleNumber(collection, feature.id, session)
+        val flags = resolveFlags(collection, session)
+        val newVersion = Version(previousMetadata.version.txn + 1)
+        val tuple = tuple(
+            session.storage,
+            tupleNumber,
+            feature,
+            metadataForNewVersion(previousMetadata, newVersion, feature, flags),
+            write.attachment,
+            flags
+        )
+
+        removeFeatureFromDel(collection, feature.id)
+        collection.history?.let { hstTable ->
+            insertPreviousVersionToHst(hstTable.name, cursorWithFeatureHead, newVersion)
+        }
+        cursorWithFeatureHead.close() // we don't need it anymore, TODO: refactor for nicer usage
+        updateFeatureInHead(collection, tuple, feature, newVersion, previousMetadata)
+        return tuple
+    }
+
+    private fun metaFromCursor(cursor: PgCursor): Metadata {
+        // TODO: maybe PgCursor.map()?
+        val txn: Int64 = cursor[PgColumn.txn]
+        return Metadata(
+            storeNumber = cursor[PgColumn.store_number],
+            updatedAt = cursor[PgColumn.updated_at],
+            createdAt = cursor[PgColumn.created_at],
+            authorTs = cursor[PgColumn.author_ts],
+            nextVersion = maybeVersion(cursor.column(PgColumn.txn_next)),
+            version = Version(txn),
+            prevVersion = maybeVersion(cursor.column(PgColumn.ptxn)),
+            uid = cursor[PgColumn.uid],
+            puid = cursor.column(PgColumn.puid) as? Int,
```

Comment:
we should se puid to 0 in case of null

Reply:
https://github.com/heremaps/naksha/pull/333#discussion_r1741733482